[DynamoFormat] Read a Set which has been deleted after an update operation

Slakah opened the pull request:
# TODO
- [ ] Fix other set types.

Fixes #244 .

regiskuckaertz commented on formats/src/main/scala/com/gu/scanamo/DynamoFormat.scala:
```diff
@@ -115,6 +115,7 @@ object DynamoFormat extends EnumDynamoFormat {
   def iso[A, B](r: B => A)(w: A => B)(implicit f: DynamoFormat[B]) = new DynamoFormat[A] {
     override def read(item: AttributeValue): Either[DynamoReadError, A] = f.read(item).map(r)
     override def write(t: A): AttributeValue = f.write(w(t))
+    override val default: Option[A] = f.default.map(r)
```
👍

regiskuckaertz commented on formats/src/main/scala/com/gu/scanamo/DynamoFormat.scala:
```diff
@@ -310,7 +311,14 @@ object DynamoFormat extends EnumDynamoFormat {
 
   private val javaNumSetFormat = attribute(_.getNS, "NS")(_.withNS)
-  private val javaStringSetFormat = attribute(_.getSS, "SS")(_.withSS)
+  private val javaStringSetFormat: DynamoFormat[java.util.List[String]] =
```
Not too keen on having instances for mutable types, but I see that was the direction chosen previously.

regiskuckaertz commented:
This is great ❤️ 

In the future, I'd prefer something like this:
```scala
private def setFormat[S, T](
  f: AttributeValue => java.util.List[S], g: java.util.List[S] => AttributeValue)(
  iso: Iso[S, T]): DynamoFormat[Set[T]]
```
and derive `stringSetFormat`, `numberSetFormat` and `binarySetFormat` from it. The `Iso` is because numbers are encoded as strings in DynamoDB and our instance wouldn't be lawful, but I can live with that.

regiskuckaertz commented:
In fact `javaNumSetFormat` is subject to the same behaviour, so will need updating too.

Slakah commented:
Sorry I should have mentioned that this is in progress. I'm going to be
moving house this next week (very busy), but should be able to progress on
the rest of this PR soon(tm).

On Tue, 28 Aug 2018, 7:31 pm Regis Kuckaertz, <notifications@github.com>
wrote:

> In fact javaNumSetFormat is subject to the same behaviour, so will need
> updating too.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scanamo/scanamo/pull/250#issuecomment-416693569>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABB1acLSeBMTY5ZRt-mAJgm1fW67sRhks5uVYybgaJpZM4WOoQT>
> .
>

Slakah commented:
Hi @regiskuckaertz 

I've updated the PR to fix the number set values as well.

I've also changed the set formatters to encode an empty set as a `null` value, due to dynamo not supporting the notion of empty sets:
> Set type attributes cannot be empty.
[source](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_PutItem.html)

This can be removed if required, as it's a tad more controversial, but it does feel like an `AttributeValue` wrote by a `DynamoFormat` should be insertable into dynamo.

wrt.
> and derive stringSetFormat, numberSetFormat and binarySetFormat from it. 

The reason this wouldn't be possible, is because the decoding of a number can fail, so at the very least an `xmap` would need to be used.

Slakah commented on formats/src/main/scala/com/gu/scanamo/DynamoFormat.scala:
```diff
@@ -309,71 +310,102 @@ object DynamoFormat extends EnumDynamoFormat {
+    new DynamoFormat[Set[T]] {
+      override def read(av: AttributeValue): Either[DynamoReadError, Set[T]] =
+        for {
+          ss <- Either.fromOption(Option(av.getSS), NoPropertyOfType("NS", av))
```
Good spot! Turns out I managed to break the tests, which was hiding the fact that I wired this in wrong. Should have been `getNS` instead of `getSS` etc.

See: https://github.com/scanamo/scanamo/pull/250/commits/bab4151900ae6c724e491e541d9df8d9d1031555

philwills commented:
Late to this, but agree with @regiskuckaertz that this looks great. Thanks @Slakah
